nifi/nifi-nar-bundles/nifi-kite-bundle/nifi-kite-processors/src/main/java/org/apache/nifi/processors/kite/KiteStorageProcessor.java

```diff
@@ -0,0 +1,169 @@
+      .name("success")
+      .description("FlowFile content has been successfully saved")
+      .build();
+
+  private static Relationship INCOMPATIBLE = new Relationship.Builder()
+      .name("incompatible")
+      .description("FlowFile content is not compatible with the target dataset")
+      .build();
+
+  private static Relationship FAILURE = new Relationship.Builder()
+      .name("failure")
+      .description("FlowFile content could not be processed")
+      .build();
+
+  public static final PropertyDescriptor KITE_DATASET_URI =
+      new PropertyDescriptor.Builder()
+          .name("Target dataset URI")
+          .description(
+              "URI that identifies a Kite dataset where data will be stored")
+          .addValidator(RECOGNIZED_URI)
+          .expressionLanguageSupported(true)
+          .required(true)
+          .build();
+
+  private static final List<PropertyDescriptor> PROPERTIES =
+      ImmutableList.<PropertyDescriptor>builder()
+          .addAll(AbstractKiteProcessor.getProperties())
+          .add(KITE_DATASET_URI)
+          .build();
+
+  private static final Set<Relationship> RELATIONSHIPS =
+      ImmutableSet.<Relationship>builder()
+          .add(SUCCESS)
+          .add(INCOMPATIBLE)
+          .add(FAILURE)
+          .build();
+
+  @Override
+  protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+    return PROPERTIES;
+  }
+
+  @Override
+  public Set<Relationship> getRelationships() {
+    return RELATIONSHIPS;
+  }
+
+  @Override
+  public void onTrigger(ProcessContext context, ProcessSession session)
+      throws ProcessException {
+    FlowFile flowFile = session.get();
+    if (flowFile == null) {
+      return;
+    }
+
+    final View<Record> target = load(context, flowFile);
+    final Schema schema = target.getDataset().getDescriptor().getSchema();
+
+    try {
+      final LongHolder writtenCount = new LongHolder(0L);
+      StopWatch timer = new StopWatch(true);
+      session.read(flowFile, new InputStreamCallback() {
+        @Override
+        public void process(InputStream in) throws IOException {
+          try (DataFileStream<Record> stream = new DataFileStream<>(
+              in, AvroUtil.newDatumReader(schema, Record.class))) {
+            IncompatibleSchemaException.check(
+                SchemaValidationUtil.canRead(stream.getSchema(), schema),
+                "Incompatible file schema %s, expected %s",
+                stream.getSchema(), schema);
+
+            long written = 0L;
+            try (DatasetWriter<Record> writer = target.newWriter()) {
+              for (Record record : stream) {
+                writer.write(record);
+                written += 1;
+              }
+            } finally {
+              writtenCount.set(written);
+            }
+          }
+        }
+      });
+      timer.stop();
+
+      session.adjustCounter("Stored records", writtenCount.get(),
```

Comment:
This needs to be in the finally block because the records are written at that point and an exception will set the count, but skip updating the counter.